### DISCUSSION TITLE
Terrain precision

### DISCUSSION THREAD

**Discussion by radoslavirha** [visgl/loaders.gl#2194] [Q&A]
Hi,

I implemented terrains in my app and everything was fine, until I added point cloud to map. Half of the model is on terrain, another half is floating (point cloud is precise and person who measured it confirmed it's ok). I'm using some government elevation data (highest precision I can get ... but I must convert it => https://github.com/visgl/loaders.gl/discussions/2193)

I compared my terrain with the same terrain in Esri viewer and it's different sometimes. Again globally it seems to be fine, we found anomaly in specific location (maybe more as we will add more models). I tried to improve my `lerc` decoding and creating `png` inspired by `terrarium`, `mapbox`, `amazon` formulas and it's always the same, so I think I'm not deforming data while converting. 

It looks different when using `delatin` and `martini` tesselator, changing `meshMaxError` in both directions does not help.

I can't identify where the issue is, but I want to ask if it can be deformed by `tesselator` somehow? Or it works only for some max precision?

Esri viewer:
<img width="1226" alt="original" src="https://user-images.githubusercontent.com/23722745/173188361-e5e48394-18e0-47b8-ac03-9b5bac84fd3d.png">

Martini in deck.gl:
<img width="1056" alt="point-cloud" src="https://user-images.githubusercontent.com/23722745/173188389-4008eeb7-a4f2-4483-9bab-358f42889489.png">

Delatin in deck.gl:
<img width="825" alt="delatin" src="https://user-images.githubusercontent.com/23722745/173188399-481d1574-a6fe-4939-bb78-0b037e30b9e2.png">

Same texture as elevation raster:
<img width="825" alt="texture" src="https://user-images.githubusercontent.com/23722745/173188411-89aaa703-6f9b-4512-9ccc-6cb7a310f143.png">

As you can see `martini` creates smooth terrain, but the slope on the left side is missing, `delatin` creates slope correctly but terrain si very deformed. I covered elevation model with the same texture and there is slope on the left side visible (blue and green color).

So what do you think? From my point of view data seems to be correct but I think something is wrong with tesselators? Yes and elevation model should be something around 20cm. Thanks

**Comment by kylebarron** [comment]
> It looks different when using `delatin` and `martini` tesselator

`delatin` and `martini` are two different algorithms. From https://github.com/mapbox/martini/issues/15#issuecomment-700475731

> Martini:
> 
> * Only works on square 2^n+1 x 2^n+1 grids.
> * Generates a hierarchy of meshes (pick arbitrary detail after a single run)
> * Optimized for meshing speed rather than quality.
> 
> Delatin:
> 
> * Works on arbitrary raster grids.
> * Generates a single mesh for a particular detail.
> * Optimized for quality (as little triangles as possible for a given error).

At a high level, Martini creates only right triangles, so the mesh will be bigger.

> changing `meshMaxError` in both directions does not help

The `meshMaxError` defines when the algorithm stops refinement of triangles. A larger `meshMaxError` leads to larger triangles that approximate the terrain less accurately.

**Reply by kylebarron** [reply]
> Half of the model is on terrain, another half is floating

In summary, this is expected because you're using two different terrain sources and because there's approximation inherent in the terrain generation. The way to "solve" it would be to "snap" point cloud data onto the terrain mesh.

**Reply by radoslavirha** [reply]
I don't want to snap point cloud to the terrain. I just realised (when I added point cloud) my terrain is deformed compared to the same terrain in ESRI viewer. On the first image there is ESRI viewer and ESRI datasource (I expect it to be good). As you can see my point cloud would fit to the terrain. Once I used the same terrain in my deck app, there is a huge gap on the left side and I don't know why. 

I don't need point cloud sitting on the terrain with cm precision, but few meters? is too much I think. Terrains are measured with some precision, this should be 20 cm if I remember, Point cloud is much more precise. It won't match.

So I'm curious why it's so different with the same elevation data and what can I do with it. :)

**Reply by kylebarron** [reply]
The default `meshMaxError` is 10 meters. So it makes sense that if you're expecting 20cm precision that the terrain will look bad. You can set `meshMaxError` to 0 but then you'll have an inefficient mesh.
https://github.com/visgl/loaders.gl/blob/d6e430d3348a87dca0cdd616f078ee980d86e6d1/modules/terrain/src/terrain-loader.ts#L19
